// ngx-logger/esm5/lib/custom-logger.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import { DatePipe } from '@angular/common';
import { NGXLoggerHttpService } from './http.service';
import { NGXLogger } from './logger.service';
import { NGXMapperService } from './mapper.service';
/**
 * CustomNGXLoggerService is designed to allow users to get a new instance of a logger
 */
var CustomNGXLoggerService = /** @class */ (function () {
    function CustomNGXLoggerService(mapperService, httpService, platformId, datePipe) {
        this.mapperService = mapperService;
        this.httpService = httpService;
        this.platformId = platformId;
        this.datePipe = datePipe;
    }
    /**
     * @param {?} config
     * @param {?=} httpService
     * @param {?=} logMonitor
     * @param {?=} mapperService
     * @return {?}
     */
    CustomNGXLoggerService.prototype.create = /**
     * @param {?} config
     * @param {?=} httpService
     * @param {?=} logMonitor
     * @param {?=} mapperService
     * @return {?}
     */
    function (config, httpService, logMonitor, mapperService) {
        // you can inject your own httpService or use the default,
        /** @type {?} */
        var logger = new NGXLogger(mapperService || this.mapperService, httpService || this.httpService, config, this.platformId, this.datePipe);
        if (logMonitor) {
            logger.registerMonitor(logMonitor);
        }
        return logger;
    };
    CustomNGXLoggerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    CustomNGXLoggerService.ctorParameters = function () { return [
        { type: NGXMapperService },
        { type: NGXLoggerHttpService },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: DatePipe }
    ]; };
    return CustomNGXLoggerService;
}());
export { CustomNGXLoggerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CustomNGXLoggerService.prototype.mapperService;
    /**
     * @type {?}
     * @private
     */
    CustomNGXLoggerService.prototype.httpService;
    /**
     * @type {?}
     * @private
     */
    CustomNGXLoggerService.prototype.platformId;
    /**
     * @type {?}
     * @private
     */
    CustomNGXLoggerService.prototype.datePipe;
}
